lib/nnc/mfa/ccv_nnc_mfa_gemm.cpp

```diff
@@ -0,0 +1,377 @@
+#include "ccv_nnc_mfa.hpp"
+#include "ccv_nnc_mfa_hash.hpp"
+#include <simd/simd.h>
+using namespace ccv::nnc;
+
+#include <string>
+
+// MARK: - C
+
+void ccv_nnc_mfa_async_prepare_gemm(mfa::context* context, ccv_nnc_mfa_gemm_params_t params)
+{
+  context->gemm_cache.prepare(context, mfa::gemm::hash(params), true);
+}
+
+void ccv_nnc_mfa_sync_prepare_gemm(mfa::context* context, ccv_nnc_mfa_gemm_params_t params)
+{
+  context->gemm_cache.prepare(context, mfa::gemm::hash(params), false);
+}
+
+void ccv_nnc_mfa_encode_gemm(mfa::context* context, ccv_nnc_mfa_gemm_params_t params, MTL::CommandBatch* command_batch, MTL::Buffer** tensors, size_t* tensor_offsets)
+{
+  mfa::gemm::hash hash(params);
+  auto iterator = context->gemm_cache.map.find(hash);
+  if (iterator == context->gemm_cache.map.end()) {
+    mfa::precondition_failure("GEMM hash not cached.", __LINE__, __FILE__, __FUNCTION__);
+  }
+  
+  auto* pipeline = iterator->second;
+  pipeline->wait();
+  
+  auto* encoder = command_batch->start_command(pipeline->get_pso());
+  encoder->setThreadgroupMemoryLength(pipeline->get_threadgroup_memory_length(), 0);
+  
+  int num_tensors = 0;
+  while (tensors[num_tensors] != nullptr) {
+    num_tensors += 1;
+  }
+  CCV_NNC_MFA_PRECONDITION(num_tensors == 3)
+  for (int i = 0; i < num_tensors; ++i) {
+    encoder->useResource(tensors[i], i < 2 ? MTL::ResourceUsageRead : MTL::ResourceUsageWrite);
```

Comment:
The specification for the buffer bindings includes more than 3 buffers. D would store biases or other user-defined metadata for fused activations. You can't assume indices 3... will be write-only.

```metal
template <typename T>
void _gemm_impl(device T *A [[buffer(0)]],
                device T *B [[buffer(1)]],
                device T *C [[buffer(2)]],
                device void *D [[buffer(3), function_constant(fused_activation)]],
```
```suggestion
    if (i < 2) {
      encoder->useResource(tensors[i], MTL::ResourceUsageRead);
    } else if (i < 3) {
      encoder->useResource(tensors[i], MTL::ResourceUsageWrite);
    } else {
      // This should never happen.
      CCV_NNC_MFA_PRECONDITION(false);
    }
```

Reply:
So you would prefer to fail on this one so later we can do some fixes to support bias integration? SGTM.

Reply:
Yes. So we don't accidentally cause a bug, if we ever do implement fused activations.